### PR TITLE
Download file with resume

### PR DESCRIPTION
**Pull request by jszlenk**
Adding a method for downloading a fragment of a file from the Arvados server, starting at a specific offset, which is useful in situations of continuing an interrupted download.

### CODE REVIEW

**Comment by tetron** [comment]
Thank you for your contribution.  Could you amend your PR to add `Arvados-DCO-1.1-Signed-off-by` to commit 6c31606c227b19d243bdb9a1c0aee7e213b1bc10 or squash it?  We have git hooks that check for `Arvados-DCO-1.1-Signed-off-by` and can't merge without it.

**Comment by jszlenk** [comment]
Sure, here you go, ready

File: sdk/java-v2/src/main/java/org/arvados/client/logic/keep/FileDownloader.java
Comment (tfmorris) [inline sdk/java-v2/src/main/java/org/arvados/client/logic/keep/FileDownloader.java:76]: It would be good to have a test for this new method.

**Review by tfmorris — CHANGES_REQUESTED** [review]
That looks like a useful addition! I've got a couple of suggestions on how it could be made more useful.  I know this enhancement is modeled on the existing API, but I think that needs to be fixed as well, because it requires the entire file to be read into memory (because it `download()` returns `byte[]`.

File: sdk/java-v2/src/main/java/org/arvados/client/logic/keep/FileDownloader.java
Comment (jszlenk) [reply sdk/java-v2/src/main/java/org/arvados/client/logic/keep/FileDownloader.java]: Done

File: sdk/java-v2/src/main/java/org/arvados/client/api/client/KeepWebApiClient.java
Comment (jszlenk) [reply sdk/java-v2/src/main/java/org/arvados/client/api/client/KeepWebApiClient.java]: Done

File: sdk/java-v2/src/main/java/org/arvados/client/logic/keep/FileDownloader.java
Comment (jszlenk) [reply sdk/java-v2/src/main/java/org/arvados/client/logic/keep/FileDownloader.java]: Done

File: sdk/java-v2/src/main/java/org/arvados/client/api/client/KeepWebApiClient.java
Comment (jszlenk) [reply sdk/java-v2/src/main/java/org/arvados/client/api/client/KeepWebApiClient.java]: Yes, I agree

**Comment by jszlenk** [comment]
I have applied the corrections, please code review

**Comment by tfmorris** [comment]
Thanks! That looks like an improvement.

Just to clarify, I'm not a member of the core team. They are the ones who will need to approve and merge this.

**Comment by brettcs** [comment]
Hi Jakub, thanks for this too. Everything looks good here process-wise. Unfortunately the test you added is failing because it assumes there's a directory named `downloads`, which isn't true on our test server. https://ci.arvados.org/job/developer-run-tests-remainder/4237/console

This test class already sets up and tears down a directory at `FILE_DOWNLOAD_TEST_DIR`. Can you please push an update to have your test use that? That should get the test working in different environments with the least effort on your part. Once tests are clean I look forward to merging. Thanks!

**Comment by jszlenk** [comment]
Sure thing. I corrected the test. Please check the code again.

**Comment by brettcs** [comment]
All this code looks good to me. Unfortunately [our Jenkins build is still failing](https://ci.arvados.org/job/developer-run-tests-remainder/4238/consoleFull):

```
org.arvados.client.logic.keep.FileDownloaderTest > testDownloadFileWithResume FAILED
    java.lang.AssertionError at FileDownloaderTest.java:139
```

This is the length equality assertion. I know our Jenkins output isn't the most helpful, so just off the top of my head, here are some things I'm wondering about to check. These are just ideas, the real problem might be something else.

* Is it possible there's some difference in file handling between Linux and other OSes?
* Does the returned file need to be seek'ed to the beginning for length to return the full result? Checking the docs that doesn't seem to be the case, but maybe it's something like this?
* Is some part of the stack adding an ending newline or other terminator to the file?

I'm happy to merge this once the tests are passing, but we do need to figure that out. Thanks again.

**Comment by jszlenk** [comment]
I made a small mistake in the test. I have corrected it.
Check again, please. Thank you for your understanding.

**Comment by brettcs** [comment]
That did it, thanks! I merged this. Thanks again, and thanks to Tom too for the suggestions, I do think this is a nicer API for folks to work with.

**Pull request merged**
